saleor/order/migrations/tasks/saleor3_13.py

```diff
@@ -46,3 +51,70 @@ def order_events_rename_transaction_capture_events_task():
     if ids:
         update_type_to_transaction_charge_requested(qs)
         order_events_rename_transaction_capture_events_task.delay()
+
+
+def quantize_price(price, currency: str):
+    precision = get_currency_precision(currency)
+    number_places = Decimal(10) ** -precision
+    return price.quantize(number_places)
+
+
+def update_order_charge_status(order: Order, granted_refund_amount: Decimal):
+    """Update the current charge status for the order.
+
+    We treat the order as overcharged when the charged amount is bigger that
+    order.total - order granted refund
+    We treat the order as fully charged when the charged amount is equal to
+    order.total - order granted refund.
+    We treat the order as partially charged when the charged amount covers only part of
+    the order.total - order granted refund
+    We treat the order as not charged when the charged amount is 0.
+    """
+    total_charged = order.total_charged_amount or Decimal("0")
+    total_charged = quantize_price(total_charged, order.currency)
+
+    current_total_gross = order.total_gross_amount - granted_refund_amount
+    current_total_gross = max(current_total_gross, Decimal("0"))
+    current_total_gross = quantize_price(current_total_gross, order.currency)
+
+    if total_charged == current_total_gross:
+        order.charge_status = OrderChargeStatus.FULL
+    elif total_charged <= Decimal(0):
+        order.charge_status = OrderChargeStatus.NONE
+    elif total_charged < current_total_gross:
+        order.charge_status = OrderChargeStatus.PARTIAL
+    else:
+        order.charge_status = OrderChargeStatus.OVERCHARGED
+
+
+@app.task
+def update_orders_charge_statuses_task(stop_granted_refund_pk=-1):
+    """Update the charge status for orders with granted refunds.
+
+    Task takes around 0.3 seconds for 1000 orders and around 5 MB of memory.
+    """
+    if stop_granted_refund_pk == -1:
```

Comment:
I can see one edge-case with this approach, e.g: we have 999 `OrderGrantedRefund`s , we go through that task with: `start_pk = -1`, this is going to work here, but we will then hit infinity loop.

So we need to add condition to not reshedule task if start_pk < 0.